[LoongArch] Allow difference across sections

MQ-mengqing opened the pull request:
For SecA != SecB but SecB is current section, fallback for pcrel{64,32} relocations. For linker relaxation being disabled and SecA == SecB, return directly for avoid record relocations. In other cases, record relocations which also allows across sections.

MQ-mengqing committed:
[LoongArch] Allow difference across sections

For SecA != SecB but SecB is current section, fallback for pcrel{64,32}
relocations. For linker relaxation being disabled and SecA == SecB,
return directly for avoid record relocations. In other cases, record
relocations which also allows across sections.

MQ-mengqing commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:
```diff
@@ -447,19 +447,16 @@ bool LoongArchAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
     if (!force) {
```
Because later it calls SA.getSection() and SB.getSection(). I want to handle the other cases by these section info.
E.g. for case (SecB == SecCur and SecA != SecCur), use Fallback() to emit PCRel.
For `llvm/test/MC/LoongArch/Relocations/sub-expr.s`,
LoongArch emits,
```
0000000000000000  000000010000006d R_LARCH_64_PCREL       0000000000000004 x + 0
0000000000000008  000000020000006d R_LARCH_64_PCREL       0000000000000004 y + 0
0000000000000010  0000000100000063 R_LARCH_32_PCREL       0000000000000004 x + 0
0000000000000014  0000000200000063 R_LARCH_32_PCREL       0000000000000004 y + 0
...
```
RISCV emits,
```
0000000000000000  0000000200000024 R_RISCV_ADD64          0000000000000004 x + 0
0000000000000000  0000000300000028 R_RISCV_SUB64          0000000000000000 .L0  + 0
0000000000000008  0000000600000024 R_RISCV_ADD64          0000000000000004 y + 0
0000000000000008  0000000500000028 R_RISCV_SUB64          0000000000000008 .L0  + 0
0000000000000010  0000000200000023 R_RISCV_ADD32          0000000000000004 x + 0
0000000000000010  0000000700000027 R_RISCV_SUB32          0000000000000010 .L0  + 0
0000000000000014  0000000600000023 R_RISCV_ADD32          0000000000000004 y + 0
0000000000000014  0000000800000027 R_RISCV_SUB32          0000000000000014 .L0  + 0
...
```

MaskRay commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:
```diff
@@ -447,19 +447,16 @@ bool LoongArchAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
-      // false and then this work can be finished by handleFixup.
-      if (&SecA != &SecB)
+      // Fallback for PCRel relocations.
+      if ((&SecA != &SecB) && (&SecB == &SecCur))
```
We don't add parentheses for `(a == b) && (c == d)`

MQ-mengqing committed:
Remove parentheses, address MaskRay's comment

MQ-mengqing committed:
Follow commit ab0931b63898 and adjust relocs of test

MaskRay commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp:
```diff
@@ -447,19 +447,16 @@ bool LoongArchAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
-      // is not same as the section of Fixup, it will report error. Just return
-      // false and then this work can be finished by handleFixup.
-      if (&SecA != &SecB)
+      // Fallback for PCRel relocations.
```
It seems that we fallback because the generic case may generate R_LARCH_32_PCREL, which is better than ADD/SUB relocation pair here? The comment should be clarified

MQ-mengqing committed:
Fix the condition of generating PCRel relocation

To handle the case of A - B which B is same section with the current,
generate PCRel relocations is better than ADD/SUB relocation pair.
We can resolve it as A - PC + PC - B. The A - PC will be resolved
as a PCRel relocation, while PC - B will serve as the addend.
If the linker relaxation is disabled, it can be done directly since
PC - B is constant. Otherwise, we should evaluate whether PC - B
is constant. If it can be resolved as PCRel, use Fallback which
generates R_LARCH_{32,64}_PCREL relocation later.

Refer to RISCV, we add isPCRelFixupResolved to evaluate whether
the PC - B is constant.

wangleiat reviewed (approved):
LGTM， thanks.
